patches/tModLoader/Terraria/Player.cs.patch

```diff
@@ -1468,6 +1476,78 @@
++				else if ((sliding || velocity.Y == 0f || flag || canJumpAgain_Cloud || canJumpAgain_Sandstorm || canJumpAgain_Blizzard || canJumpAgain_Fart || canJumpAgain_Sail || canJumpAgain_Unicorn || canJumpAgain_Santank || canJumpAgain_WallOfFleshGoat || canJumpAgain_Basilisk || ModExtraJumpLoader.AnyJumpAgain(this) || (flag2 && flag3)) && (releaseJump || (autoJump && (velocity.Y == 0f || sliding)))) {
+ 					if (mount.Active && MountID.Sets.Cart[mount.Type])
+ 						base.position.Y -= 0.001f;
+ 
+@@ -13946,7 +_,11 @@
+ 					bool flag10 = false;
+ 					bool flag11 = false;
+ 					bool flag12 = false;
++					bool cloud = false;
++					bool anyModExtraJump = false;
+ 					if (!flag2 && !flag) {
++						if (modExtraJumps.Length > 0)
++							goto jumpDecision;
+ 						if (canJumpAgain_Basilisk) {
+ 							flag12 = true;
+ 							canJumpAgain_Basilisk = false;
+@@ -13982,7 +_,11 @@
+ 						}
+ 						else {
+ 							canJumpAgain_Cloud = false;
++							cloud = true;
+ 						}
++
++						jumpDecision:
++							ModExtraJumpLoader.SetNextJump(this, ref flag12, ref flag11, ref flag10, ref flag9, ref flag5, ref flag6, ref flag7, ref flag8, ref cloud, ref anyModExtraJump);
+ 					}
+ 
+ 					canRocket = false;
+@@ -14189,7 +_,7 @@
+ 							Main.gore[num19].velocity += Main.rand.NextVector2Circular(1f, 1f) * 0.5f;
+ 						}
+ 					}
+-					else {
++					else if (cloud) {
+ 						isPerformingJump_Cloud = true;
+ 						int num20 = height;
+ 						if (gravDir == -1f)
+@@ -14214,6 +_,9 @@
+ 						Main.gore[num23].velocity.X = Main.gore[num23].velocity.X * 0.1f - velocity.X * 0.1f;
+ 						Main.gore[num23].velocity.Y = Main.gore[num23].velocity.Y * 0.1f - velocity.Y * 0.05f;
+ 					}
++					else if (anyModExtraJump) {
++						ModExtraJumpLoader.Jump(this);
++					}
+ 				}
+ 
+ 				releaseJump = false;
+@@ -14655,7 +_,7 @@
+ 
+ 		public void CarpetMovement() {
+ 			bool flag = false;
+-			if (grappling[0] == -1 && carpet && !canJumpAgain_Cloud && !canJumpAgain_Sandstorm && !canJumpAgain_Blizzard && !canJumpAgain_Fart && !canJumpAgain_Sail && !canJumpAgain_Unicorn && !canJumpAgain_Santank && !canJumpAgain_WallOfFleshGoat && !canJumpAgain_Basilisk && jump == 0 && velocity.Y != 0f && rocketTime == 0 && wingTime == 0f && !mount.Active) {
++			if (grappling[0] == -1 && carpet && !canJumpAgain_Cloud && !canJumpAgain_Sandstorm && !canJumpAgain_Blizzard && !canJumpAgain_Fart && !canJumpAgain_Sail && !canJumpAgain_Unicorn && !canJumpAgain_Santank && !canJumpAgain_WallOfFleshGoat && !canJumpAgain_Basilisk && !ModExtraJumpLoader.AnyJumpAgain(this) && jump == 0 && velocity.Y != 0f && rocketTime == 0 && wingTime == 0f && !mount.Active) {
```

Comment:
it'd be nice if at least the `canJumpAgain` logic was all handled by `ModExtraJumpLoader` via a ref property in `VanillaDoubleJump`

would make this easier to maintain

Reply:
Elaborate please, how would such an implementation look like?